### PR TITLE
Move `CREATE INDEX` optimization off the coordinator thread

### PR DESCRIPTION
**Pull request by aalexandrov**
Part of #16217 (this is part 4/5 for the "off-thread refactoring" for optimized statements).

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

This is similarly structured to #24049 and #24068. The first three commits are part of #24068 and will be merged with that PR.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

**Comment by jkosh44** [comment]
Was there a design document or design discussion around this work? Personally I think a better approach would be:

1. Extract all plan optimization out of sequencing and into a distinct phase in-between planning and sequencing (i.e. `planning -> optimization -> sequencing`).
2. Execute the optimization phase async.

I think this approach has the following benefits over the current approach:
- Allows us to move all plan optimization off thread at once instead of needing to move each plan separately.
- Simplifies the amount of internal messages that the Coordinator has to handle.
- Simplifies sequencing.

I personally find the multi-staged sequencing to be harder to follow and reason about. I made a PoC for step (1) awhile back and it seemed to work pretty well: https://github.com/MaterializeInc/materialize/pull/19034

File: src/adapter/src/coord/sequencer/inner/create_index.rs [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+    ) {
+        self.sequence_create_index_stage(
+            ctx,
+            CreateIndexStage::Validate(CreateIndexValidate { plan, resolved_ids }),
+            OpenTelemetryContext::obtain(),
+        )
+        .await;
+    }
+
+    /// Processes as many peek stages as possible.
```
Comment (jkosh44) [inline src/adapter/src/coord/sequencer/inner/create_index.rs:46]:
```suggestion
    /// Processes as many create index stages as possible.
```

File: src/adapter/src/coord/sequencer/inner/create_view.rs [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    ) {
+        self.sequence_create_view_stage(
+            ctx,
+            CreateViewStage::Validate(CreateViewValidate { plan, resolved_ids }),
+            OpenTelemetryContext::obtain(),
+        )
+        .await;
+    }
+
+    /// Processes as many peek stages as possible.
```
Comment (jkosh44) [inline src/adapter/src/coord/sequencer/inner/create_view.rs:45]:
```suggestion
    /// Processes as many create view stages as possible.
```

**Review by jkosh44 — APPROVED** [review]

**Comment by aalexandrov** [comment]
> Was there a design document or design discussion around this work?

No, this is unplanned work in order to prevent further occurrences of incidents like the one that we had last week, so I'm just following the pattern adopted by `sequence_peek`. Feel free to chat with @chaas and @benesch if you want to allocate actual time for working on that. Personally, I agree with your opinion that the code is harder to read and to modify now, but also I think that doing the ad-hoc modelling of the stages upfront will allow us to be in a better position when figuring out how to move this coordination process higher in the `Coordinator` stack.

**Pull request merged**